dvc/repo/diff.py

```diff
@@ -111,7 +111,7 @@ def _dir_output_paths(repo_tree, output):
 
     try:
         for fname in repo_tree.walk_files(output.path_info):
-            yield str(fname), repo_tree.get_hash(fname).value
+            yield str(fname), repo_tree.get_file_hash(fname).value
```

Comment:
So this enters a giant dataset and thus has to call get_file_hash for each individual file in it, which is slow, right?

Hm, feels like we could bulk this up somehow. Maybe we could call `get_hash(output.path_info)`  and then use `.dir_info`(part of HashInfo) to receive it in bulk? Clearly there are some things like the dirty state that we'll need to keep in mind, but maybe that could be a reasonable approach to save up on metadata. Classic stat&walk problem :slightly_smiling_face:

Reply:
This could work, but part of the issue right now is that dir_info/dir_cache is a list, so looking up files by the relpath field is slow

Reply:
For this diff case we could just get `.dir_info` and then iterate over that instead of calling `tree.walk()`, but for me this really feels like something that should be abstracted away inside of `walk()` (and that `get_file_hash` inside of a walk should just work without introducing any giant performance overhead)

Reply:
Really good point about the dir_info being a list! We could totally switch to dicts at least in a cached version, that will make it easier to work with everywhere. Wonder if that would mitigate the performance issue here, as it seems much more intuitive than the current two-level cache.